prov/tcp/src/tcpx_ep.c

```diff
@@ -294,21 +295,13 @@ void tcpx_ep_disable(struct tcpx_ep *ep, int cm_err)
 	case TCPX_RCVD_REQ:
 		break;
 	case TCPX_CONNECTED:
-		/* We need to remove the socket from the CQ's fdset,
-		 * or the CQ will be left in a 'signaled' state.  This
-		 * can result in threads spinning on the CQs fdset.
+		/* We need to remove the socket from the domain's fdset
+		 * or the fdset will be left in a 'signaled' state.
+		 * This can result in progress threads spinning on the fd.
 		 */
-		if (ep->util_ep.tx_cq) {
-			wait = container_of(ep->util_ep.tx_cq->wait,
-					    struct util_wait_fd, util_wait);
-			ofi_wait_fdset_del(wait, ep->bsock.sock);
-		}
-
-		if (ep->util_ep.rx_cq) {
-			wait = container_of(ep->util_ep.rx_cq->wait,
-					    struct util_wait_fd, util_wait);
-			ofi_wait_fdset_del(wait, ep->bsock.sock);
-		}
+		domain = container_of(ep->util_ep.domain, struct tcpx_domain,
+				      util_domain);
+		tcpx_unwatch_socket(domain, ep);
```

Comment:
Don't we need to remove the EP from domain->ep_list? Do we expect further events for the EP?

Reply:
I was trying to match the behavior of the current code, which only removes the socket from the pollfd.  See my other comment.  I *think* we can remove the ep from the ep_list here, but for some reason the current code does not do this.

Reply:
The tracking of an ep in the ep_list, so that we can drive progress on it, and the watching of events on the socket need to be kept separate.  The ep uses a small bounce buffer to optimize retrieving small messages from the kernel.  We could have messages in that bounce buffer when the socket is disconnected.  We remove the socket from the pollfd, so that we don't continuously trigger a POLLIN event.  But we want to continue receiving the buffered data.

Reply:
Thanks for the explanation. It makes sense to me :+1: